### PR TITLE
[XLA:GPU] turn on cuDNN fmha flag by default

### PR DESCRIPTION
**Pull request by Cjkkkk**
Follow up PR of https://github.com/openxla/xla/pull/7331 and https://github.com/openxla/xla/pull/5782 to turn on cuDNN fused attention flag by default. Use this PR to track and fix potential performance drawback or model failure if any.

### CODE REVIEW

**Review by akuegel — APPROVED** [review]

**Comment by thomasjoerg** [comment]
The Google-internal CI system complains about the following test failing:
`TritonGemmContractionDims TritonDotForceContractionDims_1_1` in `xla/service/gpu/ir_emitter_triton_test.cc:3567`


```
Value of: module->entry_computation() ->root_instruction() ->fused_instructions_computation() ->root_instruction()
Expected: an HloInstruction:
 * with opcode dot AND
 * with operand 0 which is:
     an HloInstruction outputting
       a shape:
        * with element type BF16 AND
        * with dimensions [16,32] AND
        * with
            a layout with minor to major [1,0] AND
 * with operand 1 which is:
     an HloInstruction outputting
       a shape:
        * with element type BF16 AND
        * with dimensions [40,32] AND
        * with
            a layout with minor to major [1,0] AND
 * outputting
     a shape:
      * with element type BF16 AND
      * with dimensions [16,40] AND
      * with
          a layout with minor to major [1,0]
  Actual: 0x5603378f600 (of type xla::HloInstruction*), Shape does not have dimensions [40,32]
in bf16[32,40]{1,0}
in output shape
in bitcast.5 = bf16[32,40]{1,0} bitcast(bf16[40,32]{0,1} parameter_1.1)
in operand 1
in dot.1 = bf16[16,40]{1,0} dot(bf16[16,32]{1,0} parameter_0.1, bf16[32,40]{1,0} bitcast.5), lhs_contracting_dims={1}, rhs_contracting_dims={0}
```

The test fixture is rather simple:
```
HloModule m

ENTRY e {
  parameter_0 = bf16[16,32]{1,0} parameter(0)
  parameter_1 = bf16[40,32]{0,1} parameter(1)
  ROOT dot.15148 = bf16[16,40]{1,0} dot(parameter_0, parameter_1), lhs_contracting_dims={1}, rhs_contracting_dims={1}
}
```

Would the cuDNN fMHA matcher pick up on this?

**Comment by Cjkkkk** [comment]
You can see there was a fix in https://github.com/tensorflow/tensorflow/commit/af9a94bfdbff7fe1663bf083a3133eeb00883e23 which I forgot to include. But I am not sure what caused the test to fail maybe @akuegel remember why this fix is needed in the original PR?

**Comment by akuegel** [comment]
> You can see there was a fix in https://github.com/tensorflow/tensorflow/commit/af9a94bfdbff7fe1663bf083a3133eeb00883e23 which I forgot to include. But I am not sure what caused the test to fail maybe @akuegel remember why this fix is needed in the original PR?

The new pipeline moves layout normalization towards earlier, most importantly before GemmRewriterTriton. So some triton end2end Tests are affected by this.

**Comment by Cjkkkk** [comment]
@akuegel @thomasjoerg I included the original fix for this failure and rebased on the recent merge of https://github.com/openxla/xla/pull/7331.

**Comment by Cjkkkk** [comment]
> > You can see there was a fix in [tensorflow/tensorflow@af9a94b](https://github.com/tensorflow/tensorflow/commit/af9a94bfdbff7fe1663bf083a3133eeb00883e23) which I forgot to include. But I am not sure what caused the test to fail maybe @akuegel remember why this fix is needed in the original PR?
> 
> The new pipeline moves layout normalization towards earlier, most importantly before GemmRewriterTriton. So some triton end2end Tests are affected by this.

Why would layout normalization changed the layout of dot input though?

**Comment by akuegel** [comment]
> > > You can see there was a fix in [tensorflow/tensorflow@af9a94b](https://github.com/tensorflow/tensorflow/commit/af9a94bfdbff7fe1663bf083a3133eeb00883e23) which I forgot to include. But I am not sure what caused the test to fail maybe @akuegel remember why this fix is needed in the original PR?
> > 
> > The new pipeline moves layout normalization towards earlier, most importantly before GemmRewriterTriton. So some triton end2end Tests are affected by this.
> 
> Why would layout normalization changed the layout of dot input though?

Layout assignment assigns non-default Layouts, which are then turned into default layouts which means dimensions of the shape will be permuted. If you want to see what happens step by step you can set the flag to dump hlo after every pass.

**Review by akuegel — APPROVED** [review]

**Comment by Cjkkkk** [comment]
any info on the failed internal checks?

**Comment by akuegel** [comment]
I still need to get someone to confirm that I can share the HLO, but what I noticed is that the dots have non-contiguous batch dimensions (e.g.  lhs_batch_dims={0,1,2,4}, but also happens for rhs_batch_dims). Is this potentially a problem for the custom call, and would need to be checked by the matcher?

**Comment by akuegel** [comment]
Actually, at the time when the pass is run, this has been normalized already, so this is not the problem.

**Comment by Cjkkkk** [comment]
> No engine configurations found for this opGraph and heuristics

Thanks for the info. I have not seen " No engine configurations found for this opGraph and heuristics" error before myself. A reproducer would be great!

**Comment by akuegel** [comment]
I got approval to share the HLO reproducer, when running this with run_hlo_module on A100 it produced the error.

```
HloModule reproducer

region_0.36 {
  Arg_0.37 = f16[] parameter(0)
  Arg_1.38 = f16[] parameter(1)
  ROOT maximum.39 = f16[] maximum(Arg_0.37, Arg_1.38)
}

region_1.48 {
  Arg_0.49 = f32[] parameter(0)
  Arg_1.50 = f32[] parameter(1)
  ROOT add.51 = f32[] add(Arg_0.49, Arg_1.50)
}

ENTRY main {
  arg2.3 = f16[1,8,16,5,128]{4,3,2,1,0} parameter(2), parameter_replication={false}
  reshape.14 = f16[1,8,16,5,128]{4,3,2,1,0} reshape(arg2.3)
  arg5.6 = f32[128,2,64]{2,1,0} parameter(5), parameter_replication={false}
  convert.18 = f16[128,2,64]{2,1,0} convert(arg5.6)
  dot.19 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} dot(reshape.14, convert.18), lhs_contracting_dims={4}, rhs_contracting_dims={0}, frontend_attributes={grad_x="false",grad_y="false"}
  transpose.20 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} transpose(dot.19), dimensions={0,1,2,3,4,5}
  arg6.7 = f32[2,64]{1,0} parameter(6), parameter_replication={false}
  convert.17 = f16[2,64]{1,0} convert(arg6.7)
  broadcast.21 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} broadcast(convert.17), dimensions={4,5}
  add.22 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} add(transpose.20, broadcast.21)
  arg0.1 = f16[1,8,16,5,128]{4,3,2,1,0} parameter(0), parameter_replication={false}
  reshape.12 = f16[1,8,16,5,128]{4,3,2,1,0} reshape(arg0.1)
  arg3.4 = f32[128,2,64]{2,1,0} parameter(3), parameter_replication={false}
  convert.24 = f16[128,2,64]{2,1,0} convert(arg3.4)
  dot.25 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} dot(reshape.12, convert.24), lhs_contracting_dims={4}, rhs_contracting_dims={0}, frontend_attributes={grad_x="false",grad_y="false"}
  transpose.26 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} transpose(dot.25), dimensions={0,1,2,3,4,5}
  arg4.5 = f32[2,64]{1,0} parameter(4), parameter_replication={false}
  convert.23 = f16[2,64]{1,0} convert(arg4.5)
  broadcast.27 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} broadcast(convert.23), dimensions={4,5}
  add.28 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} add(transpose.26, broadcast.27)
  constant.29 = f16[] constant(0.125)
  broadcast.30 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} broadcast(constant.29), dimensions={}
  multiply.31 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} multiply(add.28, broadcast.30)
  dot.32 = f16[1,8,16,2,5,5]{5,4,3,2,1,0} dot(add.22, multiply.31), lhs_batch_dims={0,1,2,4}, lhs_contracting_dims={5}, rhs_batch_dims={0,1,2,4}, rhs_contracting_dims={5}, frontend_attributes={grad_x="false",grad_y="false"}
  transpose.33 = f16[1,8,16,2,5,5]{4,5,3,2,1,0} transpose(dot.32), dimensions={0,1,2,3,5,4}
  constant.35 = f16[] constant(-inf)
  reduce.40 = f16[1,8,16,2,5]{4,3,2,1,0} reduce(transpose.33, constant.35), dimensions={5}, to_apply=region_0.36
  reshape.41 = f16[1,8,16,2,5,1]{5,4,3,2,1,0} reshape(reduce.40)
  broadcast.42 = f16[1,8,16,2,5,1]{5,4,3,2,1,0} broadcast(reshape.41), dimensions={0,1,2,3,4,5}
  reshape.43 = f16[1,8,16,2,5]{4,3,2,1,0} reshape(broadcast.42)
  broadcast.44 = f16[1,8,16,2,5,5]{5,4,3,2,1,0} broadcast(reshape.43), dimensions={0,1,2,3,4}
  subtract.45 = f16[1,8,16,2,5,5]{4,5,3,2,1,0} subtract(transpose.33, broadcast.44)
  exponential.46 = f16[1,8,16,2,5,5]{4,5,3,2,1,0} exponential(subtract.45)
  convert.47 = f32[1,8,16,2,5,5]{4,5,3,2,1,0} convert(exponential.46)
  constant.34 = f32[] constant(0)
  reduce.52 = f32[1,8,16,2,5]{4,3,2,1,0} reduce(convert.47, constant.34), dimensions={5}, to_apply=region_1.48
  convert.53 = f16[1,8,16,2,5]{4,3,2,1,0} convert(reduce.52)
  reshape.54 = f16[1,8,16,2,5,1]{5,4,3,2,1,0} reshape(convert.53)
  broadcast.55 = f16[1,8,16,2,5,1]{5,4,3,2,1,0} broadcast(reshape.54), dimensions={0,1,2,3,4,5}
  reshape.56 = f16[1,8,16,2,5]{4,3,2,1,0} reshape(broadcast.55)
  broadcast.57 = f16[1,8,16,2,5,5]{5,4,3,2,1,0} broadcast(reshape.56), dimensions={0,1,2,3,4}
  divide.58 = f16[1,8,16,2,5,5]{4,5,3,2,1,0} divide(exponential.46, broadcast.57)
  arg1.2 = f16[1,8,16,5,128]{4,3,2,1,0} parameter(1), parameter_replication={false}
  reshape.13 = f16[1,8,16,5,128]{4,3,2,1,0} reshape(arg1.2)
  arg7.8 = f32[128,2,64]{2,1,0} parameter(7), parameter_replication={false}
  convert.60 = f16[128,2,64]{2,1,0} convert(arg7.8)
  dot.61 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} dot(reshape.13, convert.60), lhs_contracting_dims={4}, rhs_contracting_dims={0}, frontend_attributes={grad_x="false",grad_y="false"}
  transpose.62 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} transpose(dot.61), dimensions={0,1,2,3,4,5}
  arg8.9 = f32[2,64]{1,0} parameter(8), parameter_replication={false}
  convert.59 = f16[2,64]{1,0} convert(arg8.9)
  broadcast.63 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} broadcast(convert.59), dimensions={4,5}
  add.64 = f16[1,8,16,5,2,64]{5,4,3,2,1,0} add(transpose.62, broadcast.63)
  dot.65 = f16[1,8,16,2,5,64]{5,4,3,2,1,0} dot(divide.58, add.64), lhs_batch_dims={0,1,2,3}, lhs_contracting_dims={5}, rhs_batch_dims={0,1,2,4}, rhs_contracting_dims={3}, frontend_attributes={grad_x="false",grad_y="false"}
  transpose.66 = f16[1,8,16,5,2,64]{5,3,4,2,1,0} transpose(dot.65), dimensions={0,1,2,4,3,5}
  arg9.10 = f32[2,64,128]{2,1,0} parameter(9), parameter_replication={false}
  convert.16 = f16[2,64,128]{2,1,0} convert(arg9.10)
  dot.67 = f16[1,8,16,5,128]{4,3,2,1,0} dot(transpose.66, convert.16), lhs_contracting_dims={4,5}, rhs_contracting_dims={0,1}, frontend_attributes={grad_x="false",grad_y="false"}
  transpose.68 = f16[1,8,16,5,128]{4,3,2,1,0} transpose(dot.67), dimensions={0,1,2,3,4}
  arg10.11 = f32[128]{0} parameter(10), parameter_replication={false}
  convert.15 = f16[128]{0} convert(arg10.11)
  broadcast.69 = f16[1,8,16,5,128]{4,3,2,1,0} broadcast(convert.15), dimensions={4}
  add.70 = f16[1,8,16,5,128]{4,3,2,1,0} add(transpose.68, broadcast.69)
  reshape.71 = f16[1,8,16,5,128]{4,3,2,1,0} reshape(add.70)
  tuple.72 = (f16[1,8,16,5,128]{4,3,2,1,0}) tuple(reshape.71)
  ROOT get-tuple-element.73 = f16[1,8,16,5,128]{4,3,2,1,0} get-tuple-element(tuple.72), index=0
}
```

**Comment by Cjkkkk** [comment]
I tried the HLO and seems like the fmha is called with 5D tensors while cuDNN only supports 4D tensors. Missing checks will be added.

**Comment by Cjkkkk** [comment]
@akuegel Hi Adrian, I fixed the bug where tensors with rank != 4 is passed into cuDNN and found another bug where seqlen is not divisible by 64 in training. Let me know if you want these changes in a separate PR.

**Review by akuegel — CHANGES_REQUESTED** [review]
I think landing it again with these fixes is ok. But I left two comments regarding the new tests (thanks for adding them :) )

File: xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 1]
```diff
@@ -3474,6 +3474,184 @@ ENTRY main.82 {
...
+  bitcast.120 = f16[8,16,2,5,64]{4,3,2,1,0} bitcast(add.4)
+  ROOT dot.7 = f16[8,16,2,5,64]{4,3,2,1,0} dot(bitcast.98, bitcast.120), lhs_batch_dims={0,1,2}, lhs_contracting_dims={4}, rhs_batch_dims={0,1,2}, rhs_contracting_dims={3}, frontend_attributes={grad_x="false",grad_y="false"}
+} // main
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+  CudnnFusedMHARewriter fusedMhaRewriter{
+      GetCudaComputeCapability(),
+      GetCudnnVersionWithDbiasAndMaskBwdInputSupport()};
+  TF_ASSERT_OK(RunHloPass(&fusedMhaRewriter, m.get()).status());
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter_test.cc]: Thanks for the reproducer and everything. I added the check for the status value as well.

**Review by akuegel — APPROVED** [review]

**Comment by Cjkkkk** [comment]
@akuegel hmm, seems like another google internal checks failed. Could you share some info on that?

**Comment by akuegel** [comment]
> @akuegel hmm, seems like another google internal checks failed. Could you share some info on that?

There were some infrastructure issues, but they are fixed now. I got this PR merged.

**Pull request closed without merging**